### PR TITLE
feat(checkpoint): TP recomputation communication optimization

### PR DESCRIPTION
**Pull request by li126com**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Refer to [mindspeed](https://gitee.com/ascend/MindSpeed/blob/master/docs/features/recomputation-communication.md):

When tensor parallelism is enabled, an AllReduce operation is inserted at the end of the forward pass. In the backward pass, this corresponds to an Identity operation. Since recomputation is only used to retrieve intermediate activation values, the output of the AllReduce at the end of the recomputation is redundant. Therefore, the AllReduce at the end can be removed without affecting the intermediate calculations or subsequent backward calculations.

When sequence parallelism is enabled, a ReduceScatter communication is inserted at the end of the forward pass, and an AllGather communication is inserted in the backward pass. During recomputation, the ReduceScatter communication can be directly removed. Additionally, since the gradient at the very end is not related to recomputation, the AllGather in the backward pass can overlap with the forward recomputation. This is illustrated in the diagram below.

企业微信截图_17204118124011

7B tp=4 checkpoint tgs:
mtp: 2215->2269
msp: 2309->2336

## Loss
image

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: internlm/initialize/launch.py [hunk 1]
```diff
@@ -411,6 +414,10 @@ def args_sanity_check():
             gpc.config.parallel["pipeline"].get("interleaved_overlap", False) is True
         ), "only support interleaved pipeline scheduler with overlap"
 
+    # when not use tp or sp, checkpoint_tp_no_comm should always be False
+    if gpc.config.parallel["tensor"]["size"] <= 1 and getattr(gpc.config.model, "checkpoint_tp_no_comm", False):
```
Comment (SolenoidWGT) [inline internlm/initialize/launch.py]: maybe还需要check下tpmode，如果是isp的话checkpoint_tp_no_comm应该也得为false

File: internlm/model/modeling_internlm.py [hunk 2]
```diff
@@ -179,6 +181,8 @@ def _forward(self, hidden_states, *args, **kwargs):
             cu_seqlens: 1d LongTensor, len(cu_seqlens) = hidden_states + 1
             indexes: the length of index is same as hidden states, which stand for the current position
         """
+        no_communication = args[4] if len(args) > 4 else False
```
Comment (SolenoidWGT) [inline internlm/model/modeling_internlm.py]: 用context或者函数装饰器实现no_communication的状态修改吧

File: internlm/model/modeling_internlm2.py [hunk 3]
```diff
@@ -255,7 +259,30 @@ def _dropout_and_norm_ffn(_residual, _hidden_states):
 
                     if self.residual_in_fp32:
                         residual = residual.to(torch.float32)
-                hidden_states = self.feed_forward(hidden_states)
+                hidden_states = self.feed_forward(hidden_states, no_communication=no_communication)
+
+                # pad residual
+                if no_communication and is_using_sequence_parallel() and not is_using_isp():
+                    requires_grad = residual.requires_grad
```
Comment (SolenoidWGT) [inline internlm/model/modeling_internlm2.py]: pad residual的逻辑似乎在多个文件里都有出现，考虑抽象成一个函数，减少下代码冗余？

File: internlm/solver/activation_checkpoint.py [hunk 4]
```diff
@@ -122,14 +126,34 @@ def backward(ctx, *args):
...
+        # no_communication
+        no_communication = False
+        if getattr(gpc.config.model, "checkpoint_tp_no_comm", False):
+            no_communication = True
+            inputs.append(True)
+
         detached_inputs = detach_variable(tuple(inputs))
+
+        handle = None
+        if no_communication and is_using_sequence_parallel() and not is_using_isp():
```
Comment (SolenoidWGT) [inline internlm/solver/activation_checkpoint.py]: 这个overlap优化是不是no_communication为true或者false都可以应用下？

File: internlm/model/modules/linear.py [hunk 5]
```diff
@@ -343,16 +349,16 @@ def __init__(
         else:
             super().__init__(in_features, out_features, bias=bias, device=device, dtype=dtype)
 
-    def forward(self, input: torch.Tensor) -> torch.Tensor:  # pylint: disable=W0622
+    def forward(self, input: torch.Tensor, no_communication=False) -> torch.Tensor:  # pylint: disable=W0622
```
Comment (SolenoidWGT) [inline internlm/model/modules/linear.py]: 尝试下从gpc直接获取no_communication吧，不要在函数参数列表里新增参数了

**Review by SolenoidWGT — APPROVED** [review]

File: internlm/model/modules/linear.py [hunk 6]
```diff
@@ -343,16 +349,19 @@ def __init__(
         else:
             super().__init__(in_features, out_features, bias=bias, device=device, dtype=dtype)
 
+        self.last_block_layer = False
+
     def forward(self, input: torch.Tensor) -> torch.Tensor:  # pylint: disable=W0622
         _class_name = self.__class__.__name__
         assert self._communicator is not None, f"{_class_name} should register with a communicator first."
-
+        no_communication = bool(gpc.recompute_forward_no_comm and self.last_block_layer)
```
Comment (mwiacx) [inline internlm/model/modules/linear.py]: 这个no_communiction的逻辑最好还是放在tensor.py，ops里面最好只实现和模型有关的逻辑

**Review by sunpengsdu — APPROVED** [review]

File: internlm/model/modules/linear.py [hunk 7]
```diff
@@ -589,14 +588,17 @@ def new_linear(
...
-        return RowParallelLinear(
+        linear = RowParallelLinear(
             in_features,
             out_features,
             bias,
             multiple_of,
             device,
             dtype,
         )
+        if name == "w2":
```
Comment (sunpengsdu) [inline internlm/model/modules/linear.py:599]: 这样会不会很hardcode?
